smc_pnet.8
.\" smc_pnet.8
.\"
.\"
.\" Copyright IBM Corp. 2017, 2019
.\" Author(s):  Thomas Richter <tmricht@linux.ibm.com>
.\"             Ursula Braun <ubraun@linux.ibm.com>
.\" ----------------------------------------------------------------------
.\"

.TH SMC_PNET 8 "January 2017" "smc-tools" "Linux Programmer's Manual"

.SH NAME
smc_pnet \- create, destroy, and change the SMC PNET table

.SH SYNOPSIS

.B smc_pnet
{ \fB\-a\fR | \fB\-\-add\fR \fI<pnetid>\fR }
{ \fB\-I\fR | \fB\-\-interface\fR <ethernet-interface> }
.P
.B smc_pnet
{ \fB\-a\fR | \fB\-\-add\fR \fI<pnetid>\fR }
{ \fB\-D\fR | \fB\-\-ibdevice\fR <infiniband-or-ISM-device> }
[ \fB\-P\fR | \fB\-\-ibport\fR <infiniband-port> ]
.P
.B smc_pnet
{ \fB\-a\fR | \fB\-\-add\fR \fI<pnetid>\fR }
{ \fB\-I\fR | \fB\-\-interface\fR <ethernet-interface> }
{ \fB\-D\fR | \fB\-\-ibdevice\fR <infiniband-or-ISM-device> }
[ \fB\-P\fR | \fB\-\-ibport\fR <infiniband-port> ]
.P
.B smc_pnet
{ \fB\-s\fR | \fB\-\-show\fR \fI<pnetid>\fR }
.P
.B smc_pnet
{ \fB\-d\fR | \fB\-\-delete\fR \fI<pnetid>\fR }
.P
.B smc_pnet
{ \fB\-f\fR | \fB\-\-flush\fR }
.P
.B smc_pnet
{ \fB\-v\fR | \fB\-\-version\fR }
.P
.B smc_pnet
{ \fB\-h\fR | \fB\-\-help\fR }

.SH DESCRIPTION
The SMC protocol requires grouping of standard Ethernet and RoCE networks or ISM
devices.
Such groups are called \fIPhysical Networks\fR (PNETs). The mapping is configured
within a table called \fIpnet table\fR. Any available Ethernet interface can be
combined with an available RDMA-capable network interface card (RNIC) or a
DMA-capable ISM device, if they
belong to the same Converged Ethernet fabric. To configure mapping of a RoCE Adapter
port or an ISM device to a standard Ethernet interface, both devices need to have
the same PNET ID; either hardware-defined or user-defined using the pnet table.
Hardware-defined PNET IDs cannot be overwritten.
.P
The
.B smc_pnet
command configures the pnet table.

.SH OPTIONS
By default,
.B smc_pnet
shows all entries of the pnet table.
.TP
.IR <pnetid>
defines a name for a grouping of Ethernet interface and RNICs or ISM devices.
A PNET ID consists of up to 16 alphanumeric uppercase characters without blanks.
.TP
.BR "\-a, \-\-add"
creates a new PNET ID definition to the pnet table (if it does not already exist).
Only one PNET ID can be defined for a certain
Ethernet interface, a certain InfiniBand device port or a certain ISM device.
Adding more than one PNET ID fails. Hardware defined PNET IDs cannot be overwritten.
.TP
.BR "\-s, \-\-show"
shows a certain PNET ID definition in the pnet table.
.TP
.BR "\-d, \-\-delete"
deletes an existing PNET ID definition from the pnet table.
.TP
.BR "\-f, \-\-flush"
removes all PNET ID definitions from the pnet table.
.TP
.BR "\-I, \-\-interface <ethernet-interface>"
specifies the name of the Ethernet interface to be added for a certain PNET
ID
definition.
.TP
.BR "\-D, \-\-ibdevice <infiniband-or-ISM-device>"
specifies the ID of the InfiniBand device or ISM device.
.TP
.BR "\-P, \-\-ibport <infiniband-port>"
specifies the port number of the InfiniBand device port. Valid numbers are
1 or 2.
The default value is 1.
.TP
.BR "\-v, \-\-version"
displays smc_pnet program version.
.TP
.BR "\-h, \-\-help"
displays a brief smc_pnet usage information.

.SH EXAMPLES
.B Define PNET ID ABC for the ethernet device names encf500 and bond0, and define
.B PNET ID ABC for the InfiniBand device ID 0001:00:00.0 (port 2) and the ISM
.B device ID 0004:00:00.0:
.RS 4
.PP
.nf
$ smc_pnet \-a ABC \-I encf500
$ smc_pnet \-a ABC \-I bond0
$ smc_pnet \-a ABC \-D 0001:00:00:00.0 \-P 2
$ smc_pnet \-a ABC \-D 0004:00:00:00.0
.RE
.PP
.
.B Show all pnet table entries:
.RS 4
.PP
.nf
$ smc_pnet
ABC encf500 n/a 255
ABC bond0 n/a 255
ABC n/a 0001:00:00.0 2
ABC n/a 0004:00:00.0 1
.RE
.PP
.
.B Define PNET ID XYZ for the ethernet interface name vlan0201 and the InfiniBand
.B device ID 0001:00:00.0 (port 1):
.RS 4
.PP
$ smc_pnet \-a XYZ \-I vlan0201 \-D 0001:00:00.0 \-P 1
.RE
.PP
.
.B Show all entries for PNET ID XYZ:
.RS 4
.PP
.nf
$ smc_pnet \-s XYZ
XYZ vlan0201 n/a 255
XYZ n/a 0001:00:00.0 1
.RE
.PP
.
.B Delete all pnet table entries with PNET ID named ABC:
.RS 4
.PP
$ smc_pnet \-d ABC
.RE
.PP
.
.B Delete all entries in the pnet table:
.RS 4
.PP
$ smc_pnet \-f
.RE
.PP
.
.
.SH RETURN CODES
Successful \fBsmc_pnet\fR commands return 0.
If an error occurs, \fBsmc_pnet\fR writes a message to stderr and 
completes with a return code other than 0.
.P
.SH SEE ALSO
.BR af_smc (7),
.BR smc_chk (8)
.BR smc_rnics (8),
.BR smc_run (8),
.BR smcd (8),
.BR smcr (8),
.BR smcss (8)


smc_pnet.c
/*
 * Shared Memory Communications over RDMA (SMC-R) and RoCE
 *
 * Copyright IBM Corp. 2017
 *
 * Author(s):  Thomas Richter <tmricht@linux.ibm.com>
 *
 * User space program for SMC-R PNET Table manipulation with generic netlink.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <fcntl.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/types.h>

#include <netlink/netlink.h>
#include <netlink/socket.h>
#include <netlink/msg.h>
#include <netlink/genl/genl.h>
#include <netlink/genl/ctrl.h>

#include "smctools_common.h"

static char *progname;

static struct pnetentry {
	char *pnetid;		/* Pnetid */
	char *ethname;		/* Ethernet device name */
	char *ibname;		/* Infiniband/ISM device name */
	int ibport;		/* Infiniband device port number */
	unsigned char cmd;	/* Command to execute */
} pnetcmd;

static void _usage(FILE *dest)
{
	fprintf(dest,
"Usage: %s [ OPTIONS ] [pnetid]\n"
"\t-h, --help            this message\n"
"\t-V, --version         show version information\n"
"\t-a, --add             add a pnetid entry, requires interface or ib/ism device\n"
"\t-d, --delete          delete a pnetid entry\n"
"\t-s, --show            show a pnetid entry\n"
"\t-f, --flush           flush the complete pnet table\n"
"\t-I, --interface       Ethernet interface name of a pnetid entry\n"
"\t-D, --ibdevice        Infiniband/ISM device name of a pnetid entry\n"
"\t-P, --ibport          Infiniband device port (default: 1)\n"
"\t\n"
"\tno OPTIONS            show complete pnet table\n",
		progname);
}

static void help(void) __attribute__((noreturn));
static void help(void)
{
	_usage(stdout);
	exit(EXIT_SUCCESS);
}

static void usage(void) __attribute__((noreturn));
static void usage(void)
{
	_usage(stderr);
	exit(EXIT_FAILURE);
}

static int convert(char *string)
{
	unsigned long no;
	char *endp;

	no = strtoul(string, &endp, 0);
	if (*endp != '\0' || no > 2) {
		fprintf(stderr, "%s invalid ib port:%s\n", progname, string);
		usage();
	}
	return no;
}

static const struct option long_opts[] = {
	{ "interface", 1, 0, 'I' },
	{ "ibdevice", 1, 0, 'D' },
	{ "ibport", 1, 0, 'P' },
	{ "flush", 0, 0, 'f' },
	{ "add", 0, 0, 'a'},
	{ "show", 0, 0, 's'},
	{ "delete", 0, 0, 'd'},
	{ "version", 0, 0, 'V' },
	{ "help", 0, 0, 'h' },
	{ NULL, 0, NULL, 0}
};

static struct nla_policy smc_pnet_policy[SMC_PNETID_MAX + 1] = {
	[SMC_PNETID_NAME] = {
				.type = NLA_STRING,
				.maxlen = 17
			    },
	[SMC_PNETID_ETHNAME] = {
				.type = NLA_STRING,
				.maxlen = 16
			    },
	[SMC_PNETID_IBNAME] = {
				.type = NLA_STRING,
				.maxlen = 64
			    },
	[SMC_PNETID_IBPORT] = {
				.type = NLA_U8,
				.maxlen = 1
			    }
};


/* Netlink library call back handler to be called on data reception. */
static int cb_handler(struct nl_msg *msg, void *arg)
{
	struct nlattr *attrs[SMC_PNETID_MAX + 1];
	struct nlmsghdr *hdr = nlmsg_hdr(msg);

	if (genlmsg_validate(hdr, 0, SMC_PNETID_MAX, smc_pnet_policy) ||
	   genlmsg_parse(hdr, 0, attrs, SMC_PNETID_MAX, smc_pnet_policy) < 0) {
		fprintf(stderr, "%s: invalid data returned\n", progname);
		nl_msg_dump(msg, stderr);
		return NL_STOP;
	}
	printf("%s %s %s %d\n", nla_get_string(attrs[SMC_PNETID_NAME]),
	       nla_get_string(attrs[SMC_PNETID_ETHNAME]),
	       nla_get_string(attrs[SMC_PNETID_IBNAME]),
	       nla_get_u8(attrs[SMC_PNETID_IBPORT]));
	return NL_OK;
}

static int genl_command(void)
{
	int rc = EXIT_FAILURE, id, nlmsg_flags = 0;
	struct nl_sock *sk;
	struct nl_msg *msg;

	/* Allocate a netlink socket and connect to it */
	sk = nl_socket_alloc();
	if (!sk) {
		nl_perror(NLE_NOMEM, progname);
		return rc;
	}
	rc = genl_connect(sk);
	if (rc) {
		nl_perror(rc, progname);
		rc = EXIT_FAILURE;
		goto out1;
	}
	id = genl_ctrl_resolve(sk, SMCR_GENL_FAMILY_NAME);
	if (id < 0) {
		rc = EXIT_FAILURE;
		if (id == -NLE_OBJ_NOTFOUND)
			fprintf(stderr, "%s: SMC module not loaded\n",
				progname);
		else
			nl_perror(id, progname);
		goto out2;
	}
	nl_socket_modify_cb(sk, NL_CB_VALID, NL_CB_CUSTOM, cb_handler, NULL);

	/* Allocate a netlink message and set header information. */
	msg = nlmsg_alloc();
	if (!msg) {
		nl_perror(NLE_NOMEM, progname);
		rc = EXIT_FAILURE;
		goto out2;
	}

	if ((pnetcmd.cmd == SMC_PNETID_DEL || pnetcmd.cmd == SMC_PNETID_GET) &&
	    !pnetcmd.pnetid)		/* List all */
		nlmsg_flags = NLM_F_DUMP;

	if (!genlmsg_put(msg, NL_AUTO_PORT, NL_AUTO_SEQ, id, 0, nlmsg_flags,
			 pnetcmd.cmd, SMCR_GENL_FAMILY_VERSION)) {
		nl_perror(rc, progname);
		rc = EXIT_FAILURE;
		goto out3;
	}

	switch (pnetcmd.cmd) {		/* Start message construction */
	case SMC_PNETID_ADD:
		if (pnetcmd.ethname)
			rc = nla_put_string(msg, SMC_PNETID_ETHNAME,
					    pnetcmd.ethname);
		if (rc < 0) {
			nl_perror(rc, progname);
			rc = EXIT_FAILURE;
			goto out3;
		}

		if (pnetcmd.ibname)
			rc = nla_put_string(msg, SMC_PNETID_IBNAME,
					    pnetcmd.ibname);
		if (rc < 0) {
			nl_perror(rc, progname);
			rc = EXIT_FAILURE;
			goto out3;
		}

		if (pnetcmd.ibname)
			rc = nla_put_u8(msg, SMC_PNETID_IBPORT, pnetcmd.ibport);
		if (rc < 0) {
			nl_perror(rc, progname);
			rc = EXIT_FAILURE;
			goto out3;
		}
		/* Fall through intended */
	case SMC_PNETID_DEL:
	case SMC_PNETID_GET:
		if (!pnetcmd.pnetid)		/* List all */
			break;
		rc = nla_put_string(msg, SMC_PNETID_NAME, pnetcmd.pnetid);
		if (rc < 0) {
			nl_perror(rc, progname);
			rc = EXIT_FAILURE;
			goto out3;
		}
	}

	/* Send message */
	rc = nl_send_auto(sk, msg);
	if (rc < 0) {
		nl_perror(rc, progname);
		rc = EXIT_FAILURE;
		goto out3;
	}

	/* Receive reply message, returns number of cb invocations. */
	rc = nl_recvmsgs_default(sk);
	/* Kernel commit a9d8b0b1e3d689346b016316bd91980d60c6885d
	 * introduced a misbehavior that a FLUSH of an empty table
	 * returned -ENOENT. Fix it in smc-tools as long as kernel patch did'nt
	 * land in the distros.
	 */
	if (pnetcmd.cmd == SMC_PNETID_FLUSH && rc != -NLE_OBJ_NOTFOUND)
		rc = 0;
	if (rc < 0) {
		nl_perror(rc, progname);
		rc = EXIT_FAILURE;
		goto out3;
	}
	rc = EXIT_SUCCESS;
out3:
	nlmsg_free(msg);
out2:
	nl_close(sk);
out1:
	nl_socket_free(sk);
	return rc;
}

int main(int argc, char **argv)
{
	char *slash;
	int rc, ch;

	progname = (slash = strrchr(argv[0], '/')) ? slash + 1 : argv[0];
	while ((ch = getopt_long(argc, argv, "I:D:P:fasdhvV", long_opts,
				 NULL )) != EOF) {
		switch (ch) {
		case 'f':
			if (pnetcmd.cmd)
				usage();
			pnetcmd.cmd = SMC_PNETID_FLUSH;
			break;
		case 's':
			if (pnetcmd.cmd)
				usage();
			pnetcmd.cmd = SMC_PNETID_GET;
			pnetcmd.pnetid = optarg;
			break;
		case 'd':
			if (pnetcmd.cmd)
				usage();
			pnetcmd.cmd = SMC_PNETID_DEL;
			pnetcmd.pnetid = optarg;
			break;
		case 'a':
			if (pnetcmd.cmd)
				usage();
			pnetcmd.cmd = SMC_PNETID_ADD;
			pnetcmd.pnetid = optarg;
			break;
		case 'I':
			pnetcmd.ethname = optarg;
			break;
		case 'D':
			pnetcmd.ibname = optarg;
			break;
		case 'P':
			pnetcmd.ibport = convert(optarg);
			break;
		case 'v':
		case 'V':
			printf("smc_pnet utility, smc-tools-%s (%s)\n",
			       RELEASE_STRING, RELEASE_LEVEL);
			exit(0);
		case 'h':
			help();
		case '?':
		default:
			usage();
		}
	}

	if (optind + 1 < argc) {
		fprintf(stderr, "%s too many parameters\n", progname);
		usage();
	}
	if (optind + 1 == argc)
		pnetcmd.pnetid = argv[optind];
	if (!pnetcmd.cmd) {
		if (optind < argc) {
			fprintf(stderr, "%s: parameters without option\n",
				progname);
			usage();
		}
		pnetcmd.cmd = SMC_PNETID_GET;
	}
	if (pnetcmd.cmd == SMC_PNETID_FLUSH) {
		if (optind < argc) {
			fprintf(stderr, "%s: -f takes no parameters\n",
				progname);
			usage();
		}
	}

	if (pnetcmd.cmd == SMC_PNETID_ADD) {
		if (!pnetcmd.ethname && !pnetcmd.ibname) {
			fprintf(stderr, "%s: interface or device missing\n",
				progname);
			usage();
		}
		if (!pnetcmd.ibport)
			pnetcmd.ibport = 1;
	}

	if (pnetcmd.cmd == SMC_PNETID_GET || pnetcmd.cmd == SMC_PNETID_DEL) {
		if (pnetcmd.ethname) {
			fprintf(stderr, "%s: interface %s ignored\n", progname,
					pnetcmd.ethname);
			pnetcmd.ethname = NULL;
		}
		if (pnetcmd.ibname) {
			fprintf(stderr, "%s: device %s ignored\n", progname,
					pnetcmd.ibname);
			pnetcmd.ibname = NULL;
		}
		if (pnetcmd.ibport) {
			fprintf(stderr, "%s: ibport %d ignored\n", progname,
					pnetcmd.ibport);
			pnetcmd.ibport = 0;
		}
	}
	rc = genl_command();
	return rc;
}
